code/assignTime.m

```diff
@@ -0,0 +1,281 @@
+% Exclude first packet from those to test, because need to compare times from
+% 'previous' to 'current' packets; diff_systemTick written to index of
+% second packet associated with that calculation
+numPackets = size(dataTable,1);
+for iPacket = 2:numPackets
+    diff_systemTick(iPacket,1) = mod((dataTable.systemTick(iPacket) + (2^16)...
+        - dataTable.systemTick(iPacket - 1)), 2^16);
+end
+
+% Expected elapsed time for each packet, based on sample rate and number of
+% samples per packet; in units of systemTick (1e-4 seconds)
+expectedElapsed = dataTable.packetsizes .* (1./dataTable.samplerate) * 1e4;
+
+% If diff_systemTick and expectedElapsed differ by more than 10% of expectedElapsed,
+% flag as gap
+indices_systemTickFlagged = find (abs(expectedElapsed(2:end) - diff_systemTick(2:end)) > 0.1*expectedElapsed(2:end));
+
+% All packets flagged as end of continuous chunks
+allFlaggedIndices = unique([indices_changeFs; indices_timestampFlagged;...
+    indices_dataTypeSequenceFlagged; indices_systemTickFlagged],'sorted') ;
+
+%%
+disp('Chunking data')
+% Determine indices of packets which correspond to each data chunk
+if ~isempty(allFlaggedIndices)
+    counter = 1;
+    chunkIndices = cell(1,length(allFlaggedIndices) + 1);
+    for iChunk = 1:length(allFlaggedIndices)
+        if iChunk == 1
+            chunkIndices{counter} = (1:allFlaggedIndices(1));
+            currentStartIndex = 1;
+            counter = counter + 1;
+        else
+            chunkIndices{counter} = allFlaggedIndices(currentStartIndex) + 1:allFlaggedIndices(currentStartIndex + 1);
+            currentStartIndex = currentStartIndex + 1;
+            counter = counter + 1;
+        end
+        % Last chunk, finishing with last packet
+        chunkIndices{counter} = allFlaggedIndices(currentStartIndex) + 1:numPackets;
+    end
+else
+    % No identified missing packets, all packets in one chunk
+    chunkIndices{1} = 1:numPackets;
+end
+
+%%
+% Loop through each chunk to determine offset to apply (as determined by
+% average difference between packetGenTime and expectedElapsed)
+disp('Determining start time of each chunk')
+
+% PacketGenTime in ms; convert difference to 1e-4 seconds, units of
+% systemTick and expectedElapsed
+diff_PacketGenTime = [1; diff(dataTable.PacketGenTime) * 1e1];
+
+numChunks = length(chunkIndices);
+chunksToExclude = [];
+meanError = NaN(1,numChunks);
+for iChunk = 1:numChunks
+    currentTimestampIndices = chunkIndices{iChunk};
+    
+    % Chunks must have at least 2 packets in order to have a valid
+    % diff_systemTick -- thus if chunk only one packet, it must be excluded
+    if length(currentTimestampIndices) == 1
+        chunksToExclude = [chunksToExclude iChunk];
+    end
+    % Always exclude the first packet of the chunk, because don't have an
+    % accurate diff_systemTick value for this first packet
+    currentTimestampIndices = currentTimestampIndices(2:end);
```

Comment:
I'm surprised this doesn't throw an error. Matlab has got our backs! Flagging this because we might need to add error checking for out of bound error in the future.

Reply:
Do you mean adding to make sure the length of currentTimestampIndices is >= 2?

Reply:
Sorry, I didn't mean to confuse you with this comment. To be clear, this line works perfectly, and I think there is a very very low probability this causes issues in the future. I agree with your error handling and reasoning, and this line does not need to be changed. I will mark this as a resolved conversation.  



This here is to explain the reasoning behind my earlier comment. Currently, MATLAB handles indexing out of bounds by returning an empty array. This is by design. However, should they choose to change this up for whatever reason, and move to more older style languages like C, then this line will probably break the function.